GitUI/UserControls/WarningHeader.cs

```diff
@@ -0,0 +1,193 @@
+﻿using System;
+using System.Threading.Tasks;
+using System.Windows.Forms;
+using GitCommands;
+using GitUI.CommandsDialogs;
+using Microsoft.VisualStudio.Threading;
+
+namespace GitUI.UserControls
+{
+    public partial class WarningHeader : UserControl
+    {
+        private enum State
+        {
+            None,
+            Bisect,
+            Rebase,
+            Merge,
+            Patch
+        }
+
+        private State _state = State.None;
+        private bool _hasConflicts = false;
+
+        public event EventHandler BisectEvent;
+
+        public WarningHeader()
+        {
+            InitializeComponent();
+        }
+
+        public async Task RefreshAsync(GitCommands.GitModule module)
+        {
+            // get the current state of the repo
+            await TaskScheduler.Default;
+
+            if (module.IsValidGitWorkingDir())
+            {
+                if (module.InTheMiddleOfBisect())
+                {
+                    await SetStateAsync(State.Bisect, false);
+                }
+                else
+                {
+                    bool conflicted = module.InTheMiddleOfConflictedMerge();
+
+                    if (module.InTheMiddleOfRebase())
+                    {
+                        await SetStateAsync(State.Rebase, conflicted);
+                    }
+                    else if (module.InTheMiddleOfMerge())
+                    {
+                        await SetStateAsync(State.Merge, conflicted);
+                    }
+                    else if (module.InTheMiddleOfPatch())
+                    {
+                        await SetStateAsync(State.Patch, conflicted);
+                    }
+                    else
+                    {
+                        await SetStateAsync(State.None, false);
+                    }
+                }
+            }
+
+            async Task SetStateAsync(State state, bool conflicted)
```

Comment:
Please make it an instance method, and have it tested, exposing the method via `TestAccessor`.

Reply:
I'm having a little trouble with this. I wonder if you could give any guidance on how best to expose the State enum, and this async function for testing?

I presume I would then test all properties that I set such as _status, _hasConflicts, BackColour, ButtonContainer.Controls and Label.Text?

Reply:
Currently, it's a nested function (even a closure).
Make it an instance method, i.e. factor it out as a private member function of this class.
Then apply the `TestAccessor` pattern like e.g. at the bottom of `FormCommit.cs`.
The visibility level of the `enum` can be raised to `internal`.